README.md
# nginx_server Cookbook
[![Build Status](https://travis-ci.org/kemra102/nginx_server-cookbook.svg?branch=master)](https://travis-ci.org/kemra102/nginx_server-cookbook)

#### Table of Contents

1. [Overview](#overview)
2. [Requirements](#requirements)
3. [Attributes](#attributes)
4. [Usage](#usage)
    * [nginx_server_vhost](#nginx_server_vhost)
    * [nginx_server_upstream](#nginx_server_upstream)
5. [Contributing](#contributing)
6. [License & Authors](#license-and-authors)

## Overview

This module manages the installation and configuration of NGINX.

## Requirements

Requires Chef 12.5 or later as this cookbook makes use of [Custom Resources](https://www.chef.io/blog/2015/10/08/chef-client-12-5-released/).

## Attributes

### nginx_server::default
| Key                               | Type      | Description                                   | Default |
|:---------------------------------:|:---------:|:---------------------------------------------:|:-------:|
| `['nginx_server']['manage_repo']` | `Boolean` | If the cookbook should manage the NGINX repo. | `true`  |
| `['nginx_server']['repo']` | `String` | Which repo to install NGINX from. Only used when `['nginx_server']['manage_repo']` is `true`. Possible values are; `epel`, `nginx-stable` & `nginx-mainline`. | `nginx-stable`  |
| `['nginx_server']['manage_confd']` | `Boolean` | Whether or not we should *zap* `/etc/nginx/conf.d`. | `true`  |
| `['nginx_server']['error_log_level']` | `String` | [http://nginx.org/en/docs/ngx_core_module.html#error_log](http://nginx.org/en/docs/ngx_core_module.html#error_log) | `warn`  |
| `['nginx_server']['config']['worker_connections']` | `Integer` | [http://nginx.org/en/docs/ngx_core_module.html#worker_connections](http://nginx.org/en/docs/ngx_core_module.html#worker_connections) | `1024`  |
| `['nginx_server']['config']['log_format']` | `String` | [http://nginx.org/en/docs/http/ngx_http_log_module.html#log_format](http://nginx.org/en/docs/http/ngx_http_log_module.html#log_format) | ```"\'$remote_addr - \$remote_user [\$time_local] \"\$request\" ' '\$status \$body_bytes_sent \"\$http_referer\" ' '\"\$http_user_agent\" \"\$http_x_forwarded_for\"'"``` |
| `['nginx_server']['config']['sendfile']` | `String` | [http://nginx.org/en/docs/http/ngx_http_core_module.html#sendfile](http://nginx.org/en/docs/http/ngx_http_core_module.html#sendfile) | `on`  |
| `['nginx_server']['config']['tcp_nopush']` | `String` | [http://nginx.org/en/docs/http/ngx_http_core_module.html#tcp_nopush](http://nginx.org/en/docs/http/ngx_http_core_module.html#tcp_nopush) | `off`  |
| `['nginx_server']['config']['keepalive_timeout']` | `Integer` | [http://nginx.org/en/docs/http/ngx_http_core_module.html#keepalive_timeout](http://nginx.org/en/docs/http/ngx_http_core_module.html#keepalive_timeout) | `65`  |
| `['nginx_server']['config']['gzip']` | `String` | [http://nginx.org/en/docs/http/ngx_http_gzip_module.html#gzip](http://nginx.org/en/docs/http/ngx_http_gzip_module.html#gzip) | `off`  |
| `['nginx_server']['config']['additional']` | `Hash` | Additional config to be applied to the main NGINX config. | `N/A` |

## Usage

You always need to include the main recipe:

```ruby
include_recipe 'nginx_server::default'
```

This recipe:

* Sets up the NGINX repo (**stable** repo by default).
* Installs the `nginx` package.
* Writes the main config to `/etc/nginx/nginx.conf`.
* *zaps* any files in `/etc/nginx/conf.d/` not managed by Chef.
* Enables & starts the `nginx` service.

### nginx_server_vhost

This resource defines a standard NGINX vhost.

Each `nginx_server_vhost` has the following attributes:

| Attribute     | Type                 | Description                                                                  | Default                 |
|:-------------:|:--------------------:|:----------------------------------------------------------------------------:|:-----------------------:|
| `name`        | `String` or `Symbol` | Resource name.                                                               | `N/A`                   |
| `listen`      | `Array`              | An array of hashes that describes the NGINX listen statements for the vhost. | `N/A`                   |
| `server_name` | `String` or `Array`  | Server name(s) that the vhost should respond to.                             | `N/A`                   |
| `root`        | `String`             | Web root of the vhost.                                                       | `/usr/share/nginx/html` |
| `index`       | `String`             | Index file for the website.                                                  | `index.html`            |
| `config`      | `Hash`               | Additional config options to pass to the vhost.                              | `N/A`                   |

To setup a basic vhost for example:

```ruby
nginx_server_vhost 'example.org' do
  server_name [
    'example.org',
    'www.example.org'
  ]
  root '/var/www/example.org'
  action :create
end
```

#### `listen`

The `listen` statement is an array of hashes that consists of:

* `ipaddress`
* `port`
* `options`

All parts of each hash are options and have sane defaults:

* `ipaddress` = `0.0.0.0`
* `port` = `80`
* `options` = `N/A`

For example to listen on the loopback address with SSL:

```ruby
listen [
  {
    'ipaddress': '127.0.0.1',
    'port': 443,
    'options': 'ssl'
  }
]
```

this produces a line like this:

```sh
listen 127.0.0.1:443 ssl;
```

For more examples see the cookbook's integration test cookbook.

#### `config`

The `config` is a hash of additional config for the vhost not provided by the other properties.

This supports nested config such as `location` statements as well, for example:

```ruby
config ({
    'error_page' => '404 /404.html',
    'location ~ \.php$' => {
      'try_files' => '$uri =404',
      'fastcgi_pass' => 'unix:/var/run/php5-fpm.sock',
      'fastcgi_index' => 'index.php',
      'fastcgi_param' => 'SCRIPT_FILENAME $document_root$fastcgi_script_name',
      'include' => 'fastcgi_params'
    }
  })
```

this is rendered as:

```sh
error_page 404 /404.html;
location ~ \.php$ {
  try_files $uri =404;
  fastcgi_pass unix:/var/run/php5-fpm.sock;
  fastcgi_index index.php;
  fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
  include fastcgi_params;
}
```

For more examples see the cookbook's integration test cookbook.

##### `fastcgi_param`

Because `config` is declared as a Ruby Hash you cannot declare multiple instances of `fastcgi_param` as each new declaration will override the previous.

In order to work around this you can specify multiple instances of `fastcgi_param` using `fastcgi_params`:

```ruby
config ({
  'location ~* \.php$' => {
    'fastcgi_pass' => 'unix:/var/run/php5-fpm.sock',
    'fastcgi_index' => 'index.php',
    'include' => 'fastcgi_params',
    'fastcgi_params' => [
      'SCRIPT_FILENAME $document_root$fastcgi_script_name',
      'SCRIPT_NAME $fastcgi_script_name',
      'SERVER_NAME $host'
    ]
  }
})
```

### nginx_server_upstream

The `nginx_server_upstream` defines an NGINX upstream to be used in tandem with a `nginx_server_vhost` that uses `proxy_pass` or similar to proxy connections to backend servers.

Each `nginx_server_upstream` has the following attributes:

| Attribute       | Type                 | Description                                                                                      | Default |
|:---------------:|:--------------------:|:------------------------------------------------------------------------------------------------:|:--------|
| `name`          | `String` or `Symbol` | Resource name.                                                                                   | `N/A`   |
| `upstream_name` | `String`             | Name of the upstream. Defaults to `name` if not set.                                             | `N/A`   |
| `servers`       | `Array`              | An Array of Arrays of backend servers that the upstream proxies to including associated options. | `N/A`   |
| `ip_hash`       | `Boolean`            | Whether or not to turn on the `ip_hash` functionality.                                           | `false` |
| `keepalive`     | `Integer`            | If set to an Integer turn on and set the `keepalive` functionality with the value supplied.      | `N/A`   |
| `least_conn`    | `Boolean`            | Whether or not to turn on the `least_conn` functionality.                                        | `false` |

To proxy through to a group of application servers for example:

```ruby
nginx_server_upstream 'appservers' do
  ip_hash true
  least_conn true
  servers [
    ['192.168.0.100:8080', 'max_fails=3', 'fail_timeout=30s'],
    ['192.168.0.101:8080', 'max_fails=3', 'fail_timeout=30s'],
    ['192.168.0.102:8080', 'max_fails=3', 'fail_timeout=30s']
  ]
  action :create
end
```

will render as:

```sh
upstream appservers {
  ip_hash;
  least_conn;

  server 192.168.0.100:8080 max_fails=3 fail_timeout=30s;
  server 192.168.0.101:8080 max_fails=3 fail_timeout=30s;
  server 192.168.0.102:8080 max_fails=3 fail_timeout=30s;
}
```

## Contributing

If you would like to contribute to this cookbook please follow these steps;

1. Fork the repository on Github.
2. Create a named feature branch (like `add_component_x`).
3. Write your change.
4. Write tests for your change (if applicable).
5. Write documentation for your change (if applicable).
6. Run the tests, ensuring they all pass.
7. Submit a Pull Request using GitHub.

## License and Authors

License: [BSD 2-clause](https://tldrlegal.com/license/bsd-2-clause-license-\(freebsd\)

Authors:

  * [Danny Roberts](https://github.com/kemra102)
  * [All Contributors](https://github.com/kemra102/yumserver-cookbook/graphs/contributors)


CHANGELOG.md
## 2016-12-29 - Release 1.2.0
### Summary
Small feature release.

#### Features
- Fixed [Issue #2](https://github.com/kemra102/nginx_server-cookbook/issues/2): Added support to vhost template to enable multiple instances of `fastcgi_param`.

## 2016-10-14
### Summary
Small feature release

#### Bugfixes
- Reload instead of restart when updating the main config in order to avoid interruptions to service.

#### Features
- Can now add additional parameters to the main NGINX config.

## 2016-02-09
### Small bugfix release

#### Bugfixes
- `nginx_server_vhost` now correctly takes an Array of `server_name` values.

## 2016-01-26
### Summary
Initial release.

#### Features
- Optionally manage repo to install NGINX from.
- Install NGINX.
- Manage main configuration.
- Add vhost and upstream configuration via custom resources.
- *zap* config files not managed by Chef.
- Manage the NGINX service.


test/integration/default/serverspec/nginx_spec.rb
require 'serverspec'

set :backend, :exec

describe file('/etc/yum.repos.d/nginx-mainline.repo') do
  it { should be_file }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
  it { should be_mode 644 }
end

describe package('nginx') do
  it { should be_installed.with_version('1.11') }
end

describe file('/etc/nginx/nginx.conf') do
  it { should be_file }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
  it { should be_mode 644 }
  it { should contain 'error_log /var/log/nginx/error.log warn;' }
  it { should contain 'worker_connections 1024;' }
  it { should contain 'log_format main \'$remote_addr - $remote_user [$time_local] "$request" \'' }
  it { should contain 'sendfile   on;' }
  it { should contain 'tcp_nopush off;' }
  it { should contain 'keepalive_timeout 65;' }
  it { should contain 'gzip off;' }
  it { should contain 'gzip_http_version 1.0;' }
  it { should contain 'gzip_comp_level 5;' }
  it { should contain 'gzip_proxied any;' }
  it { should contain 'gzip_vary on;' }
  it { should contain 'gzip_types application/atom+xml application/javascript application/json application/ld+json application/manifest+json application/rss+xml application/vnd.geo+json application/vnd.ms-fontobject application/x-font-ttf application/x-web-app-manifest+json application/xhtml+xml application/xml font/opentype image/bmp image/svg+xml image/x-icon text/cache-manifest text/css text/plain text/vcard text/vnd.rim.location.xloc text/vtt text/x-component text/x-cross-domain-policy text/html;' }
end

describe file('/etc/nginx/conf.d/default.conf') do
  it { should_not exist }
end

describe file('/etc/nginx/conf.d/example_ssl.conf') do
  it { should_not exist }
end

describe file('/etc/nginx/conf.d/minimal_config.conf') do
  it { should be_file }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
  it { should be_mode 644 }
  it { should contain 'root /usr/share/nginx/html;' }
  it { should contain 'index index.html;' }
  it { should contain 'server_name www.example.org;' }
end

describe file('/etc/nginx/conf.d/kitchen_sink.conf') do
  it { should contain 'server_name www.example.org example.org;' }
  it { should contain 'listen 0.0.0.0:81 spdy;' }
  it { should contain 'listen *:82 spdy;' }
  it { should contain 'listen 0.0.0.0:83 spdy;' }
  it { should contain 'listen *:443 ssl spdy;' }
  it { should contain 'error_page 404 /404.html;' }
  it { should contain 'location ~ \.php$ {' }
  it { should contain 'try_files $uri =404;' }
  it { should contain 'fastcgi_pass unix:/var/run/php5-fpm.sock;' }
  it { should contain 'fastcgi_index index.php;' }
  it { should contain 'fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;' }
  it { should contain 'include fastcgi_params;' }
end

describe file('/etc/nginx/conf.d/fastcgi_params.conf') do
  it { should contain 'location ~* \.php$ {' }
  it { should contain 'fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;' }
  it { should contain 'fastcgi_param SCRIPT_NAME $fastcgi_script_name;' }
  it { should contain 'fastcgi_param SERVER_NAME $host;' }
end

describe file('/etc/nginx/conf.d/simple_upstream.conf') do
  it { should be_file }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
  it { should be_mode 644 }
  it { should contain 'upstream simple {' }
  it { should contain 'server 127.0.0.1:8080 ;' }
end

describe file('/etc/nginx/conf.d/load_balancer.conf') do
  it { should contain 'upstream webservers {' }
  it { should contain 'server 192.168.0.100 max_fails=3 fail_timeout=30s;' }
  it { should contain 'server 192.168.0.101 max_fails=3 fail_timeout=30s;' }
  it { should contain 'server 192.168.0.102 max_fails=3 fail_timeout=30s;' }
end

describe file('/etc/nginx/conf.d/all_options.conf') do
  it { should contain 'upstream appservers {' }
  it { should contain 'server 192.168.1.100:8000 max_fails=3 fail_timeout=30s;' }
  it { should contain 'server 192.168.1.101:8000 max_fails=3 fail_timeout=30s;' }
  it { should contain 'server 192.168.1.102:8000 max_fails=3 fail_timeout=30s;' }
  it { should contain 'ip_hash;' }
  it { should contain 'keepalive 30;' }
  it { should contain 'least_conn;' }
end

describe port(80) do
  it { should be_listening.on('0.0.0.0').with('tcp') }
end

describe port(81) do
  it { should be_listening.on('0.0.0.0').with('tcp') }
end

describe port(82) do
  it { should be_listening.on('0.0.0.0').with('tcp') }
end

describe port(83) do
  it { should be_listening.on('0.0.0.0').with('tcp') }
end

describe port(443) do
  it { should be_listening.on('0.0.0.0').with('tcp') }
end

describe service('nginx') do
  it { should be_enabled }
  it { should be_running }
end


test/integration/minimal/serverspec/nginx_spec.rb
require 'serverspec'

set :backend, :exec

describe file('/etc/yum.repos.d/nginx-stable.repo') do
  it { should be_file }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
  it { should be_mode 644 }
end

describe package('nginx') do
  it { should be_installed }
end

describe file('/etc/nginx/nginx.conf') do
  it { should be_file }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
  it { should be_mode 644 }
  it { should contain 'error_log /var/log/nginx/error.log warn;' }
  it { should contain 'worker_connections 1024;' }
  it { should contain 'log_format main \'$remote_addr - $remote_user [$time_local] "$request" \'' }
  it { should contain 'sendfile   on;' }
  it { should contain 'tcp_nopush off;' }
  it { should contain 'keepalive_timeout 65;' }
  it { should contain 'gzip off;' }
end

describe file('/etc/nginx/conf.d/default.conf') do
  it { should_not exist }
end

describe file('/etc/nginx/conf.d/example_ssl.conf') do
  it { should_not exist }
end

describe service('nginx') do
  it { should be_enabled }
  it { should be_running }
end

describe port(80) do
  it { should be_listening }
end

describe file('/etc/nginx/conf.d/minimal_config.conf') do
  it { should be_file }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
  it { should be_mode 644 }
  it { should contain 'root /usr/share/nginx/html;' }
  it { should contain 'index index.html;' }
  it { should contain 'server_name www.example.org;' }
end


test/integration/upstream/serverspec/nginx_spec.rb
require 'serverspec'

set :backend, :exec

describe file('/etc/nginx/conf.d/simple_upstream.conf') do
  it { should be_file }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
  it { should be_mode 644 }
  it { should contain 'upstream simple {' }
  it { should contain 'server 127.0.0.1:8080 ;' }
end

describe file('/etc/nginx/conf.d/load_balancer.conf') do
  it { should contain 'upstream webservers {' }
  it { should contain 'server 192.168.0.100 max_fails=3 fail_timeout=30s;' }
  it { should contain 'server 192.168.0.101 max_fails=3 fail_timeout=30s;' }
  it { should contain 'server 192.168.0.102 max_fails=3 fail_timeout=30s;' }
end

describe file('/etc/nginx/conf.d/all_options.conf') do
  it { should contain 'upstream appservers {' }
  it { should contain 'server 192.168.1.100:8000 max_fails=3 fail_timeout=30s;' } # rubocop:disable Metrics/LineLength
  it { should contain 'server 192.168.1.101:8000 max_fails=3 fail_timeout=30s;' } # rubocop:disable Metrics/LineLength
  it { should contain 'server 192.168.1.102:8000 max_fails=3 fail_timeout=30s;' } # rubocop:disable Metrics/LineLength
  it { should contain 'ip_hash;' }
  it { should contain 'keepalive 30;' }
  it { should contain 'ntlm;' }
  it { should contain 'least_conn;' }
end


recipes/_repo.rb
unless node.platform_family == 'rhel' || node.platform_family == 'debian'
  Chef::Application.fatal!("Unsupported Linux distribution - #{node.platform}", 1)
end

case node['nginx_server']['repo']
when 'epel' && node.platform_family?('rhel')
  include_recipe 'yum-epel::default'
when 'nginx-stable'
  if node.platform_family?('rhel')
    include_recipe 'yum-nginx::default'
  else
    include_recipe 'apt-nginx::default'
  end
when 'nginx-mainline'
  if node.platform_family?('rhel')
    node.default['yum-nginx']['repos']['nginx-stable']['managed'] = false
    node.default['yum-nginx']['repos']['nginx-mainline']['managed'] = true
    include_recipe 'yum-nginx::default'
  else
    node.default['apt-nginx']['repos']['nginx-stable']['managed'] = false
    node.default['apt-nginx']['repos']['nginx-mainline']['managed'] = true
    include_recipe 'apt-nginx::default'
  end
else
  Chef::Application.fatal!("#{node['nginx_server']['repo']} is not an allowed
    value for node['nginx_server']['repo'].", 1)
end


recipes/default.rb
#
# Cookbook Name:: nginx_server
# Recipe:: default
#
# Copyright 2016, Danny Roberts
#
# BSD-2-Clause
#
include_recipe 'nginx_server::_repo' if node['nginx_server']['manage_repo']

package 'nginx'

template '/etc/nginx/nginx.conf' do
  cookbook 'nginx_server'
  source 'nginx.conf.erb'
  owner 'root'
  group 'root'
  mode '0644'
  verify 'nginx -t -c /etc/nginx/nginx.conf'
  notifies :reload, 'service[nginx]', :delayed
end

if node['nginx_server']['manage_confd'] # ~FC023
  zap_directory '/etc/nginx/conf.d' do
    klass [Chef::Resource::File, Chef::Resource::Template,
           Chef::Resource::Link, Chef::Resource::NginxServerVhost,
           Chef::Resource::NginxServerUpstream]
  end
end

service 'nginx' do
  action [:enable, :start]
  supports restart: true, reload: true, status: true
end


resources/block.rb
property :name, [String, Symbol], required: true, name_property: true
property :listen, Array, required: false
property :server_name, [String, Array], required: false
property :root, String, required: true, default: '/usr/share/nginx/html'
property :index, String, required: true, default: 'index.html'
property :config, Hash, required: false

def real_server_name
  server_name || name
end

def real_listen # rubocop:disable Metrics/MethodLength
  real_listen = []
  return real_listen if listen.nil?
  listen.each do |line|
    ipaddress = line['ipaddress'] || '*'
    port = line['port'] || 80
    if line['options'].is_a?(Array)
      options = line['options'].join(' ')
    else
      options = line['options'] || ''
    end
    real_listen << "listen #{ipaddress}:#{port} #{options};"
  end
  real_listen
end

def path
  "/etc/nginx/conf.d/#{name.tr(' ', '_')}.conf"
end

action :create do
  global_nginx = resources('service[nginx]')

  template path do
    cookbook 'nginx_server'
    source 'server_block.conf.erb'
    owner 'root'
    group 'root'
    mode '0644'
    variables(
      listen: real_listen,
      server_name: real_server_name,
      root: root,
      index: index,
      config: config
    )
    notifies :reload, global_nginx, :delayed
  end
end

action :delete do
  global_nginx = resources('service[nginx]')

  template path do
    action :delete
    notifies :reload, global_nginx, :delayed
  end
end


resources/vhost.rb
property :name, [String, Symbol], required: true, name_property: true
property :listen, Array, required: false
property :server_name, [String, Array], required: false
property :root, String, required: true, default: '/usr/share/nginx/html'
property :index, String, required: true, default: 'index.html'
property :config, Hash, required: false

default_action :create

def real_server_name
  server_name || name
end

def real_listen # rubocop:disable Metrics/MethodLength
  real_listen = []
  return real_listen if listen.nil?
  listen.each do |line|
    ipaddress = line['ipaddress'] || '*'
    port = line['port'] || 80
    if line['options'].is_a?(Array)
      options = line['options'].join(' ')
    else
      options = line['options'] || ''
    end
    real_listen << "listen #{ipaddress}:#{port} #{options};"
  end
  real_listen
end

def path
  "/etc/nginx/conf.d/#{name.tr(' ', '_')}.conf"
end

action :create do
  global_nginx = resources('service[nginx]')

  template path do
    cookbook 'nginx_server'
    source 'vhost.conf.erb'
    owner 'root'
    group 'root'
    mode '0644'
    variables(
      listen: real_listen,
      server_name: real_server_name,
      root: root,
      index: index,
      config: config
    )
    notifies :reload, global_nginx, :delayed
  end
end

action :delete do
  global_nginx = resources('service[nginx]')

  file path do
    action :delete
    notifies :reload, global_nginx, :delayed
  end
end


resources/upstream.rb
property :name, [String, Symbol], required: true, name_property: true
property :upstream_name, String, required: false
property :servers, Array, required: true
property :ip_hash, [FalseClass, TrueClass], required: false, default: false
property :keepalive, Integer, required: false
property :least_conn, [FalseClass, TrueClass], required: false, default: false

default_action :create

def real_upstream_name
  upstream_name || name
end

def path
  "/etc/nginx/conf.d/#{name.tr(' ', '_')}.conf"
end

action :create do
  global_nginx = resources('service[nginx]')

  template path do
    cookbook 'nginx_server'
    source 'upstream.conf.erb'
    owner 'root'
    group 'root'
    mode '0644'
    variables(
      upstream_name: real_upstream_name,
      servers: servers,
      ip_hash: ip_hash,
      keepalive: keepalive,
      least_conn: least_conn
    )
    notifies :reload, global_nginx, :delayed
  end
end

action :delete do
  global_nginx = resources('service[nginx]')

  file path do
    action :delete
    notifies :reload, global_nginx, :delayed
  end
end


resources/nginx_vhost.rb
resource_name :nginx_vhost

property :name, [String, Symbol], required: true, name_property: true
property :listen, Array, required: false
property :server_name, [String, Array], required: false
property :root, String, required: true, default: '/usr/share/nginx/html'
property :index, String, required: true, default: 'index.html'
property :config, Hash, required: false

default_action :create

def real_server_name
  server_name || name
end

def real_listen # rubocop:disable Metrics/MethodLength
  real_listen = []
  return real_listen if listen.nil?
  listen.each do |line|
    ipaddress = line['ipaddress'] || '*'
    port = line['port'] || 80
    if line['options'].is_a?(Array)
      options = line['options'].join(' ')
    else
      options = line['options'] || ''
    end
    real_listen << "listen #{ipaddress}:#{port} #{options};"
  end
  real_listen
end

def path
  "/etc/nginx/conf.d/#{name.tr(' ', '_')}.conf"
end

action :create do
  global_nginx = resources('service[nginx]')

  template path do
    cookbook 'nginx_server'
    source 'vhost.conf.erb'
    owner 'root'
    group 'root'
    mode '0644'
    variables(
      listen: real_listen,
      server_name: real_server_name,
      root: root,
      index: index,
      config: config
    )
    notifies :reload, global_nginx, :delayed
  end
end

action :delete do
  global_nginx = resources('service[nginx]')

  file path do
    action :delete
    notifies :reload, global_nginx, :delayed
  end
end
